modules/imgcodecs/src/loadsave.cpp

```diff
@@ -1121,27 +1121,43 @@ bool imdecodemulti(InputArray _buf, int flags, CV_OUT std::vector<Mat>& mats, co
     }
 }
 
-bool imencode( const String& ext, InputArray _image,
+bool imencode( const String& ext, InputArray _img,
                std::vector<uchar>& buf, const std::vector<int>& params_ )
 {
     CV_TRACE_FUNCTION();
 
-    Mat image = _image.getMat();
-    CV_Assert(!image.empty());
-
-    int channels = image.channels();
-    CV_Assert( channels == 1 || channels == 3 || channels == 4 );
-
     ImageEncoder encoder = findEncoder( ext );
     if( !encoder )
         CV_Error( Error::StsError, "could not find encoder for the specified extension" );
 
-    if( !encoder->isFormatSupported(image.depth()) )
+    std::vector<Mat> img_vec;
+    CV_Assert(!_img.empty());
+    if (_img.isMatVector() || _img.isUMatVector())
+        _img.getMatVector(img_vec);
+    else
+        img_vec.push_back(_img.getMat());
+
+    CV_Assert(!img_vec.empty());
+    const bool isMultiImg = img_vec.size() > 1;
+
+    std::vector<Mat> write_vec;
+    for (size_t page = 0; page < img_vec.size(); page++)
     {
-        CV_Assert( encoder->isFormatSupported(CV_8U) );
+        Mat image = img_vec[page];
+        CV_Assert(!image.empty());
+
+        const int channels = image.channels();
+        CV_Assert( channels == 1 || channels == 3 || channels == 4 );
+
         Mat temp;
-        image.convertTo(temp, CV_8U);
-        image = temp;
+        if( !encoder->isFormatSupported(image.depth()) )
+        {
+            CV_Assert( encoder->isFormatSupported(CV_8U) );
```

Comment:
i think all encoders supports CV_8U so imho no need to this check

Reply:
> i think all encoders supports CV_8U so imho no need to this check

Currently almost encoders supports CV_8U, but OpenEXR supports only CV_32F. And I think keeping behaviour between `imwrite()` and `imencode()` makes easy to debug encoders for current and future.

https://docs.opencv.org/4.10.0/d4/da8/group__imgcodecs.html#ga8ac397bd09e48851665edbe12aa28f25

> With OpenEXR encoder, only 32-bit float (CV_32F) images can be saved.
> 8-bit unsigned (CV_8U) images are not supported.

https://github.com/opencv/opencv/blob/450e741f8d53ff12b4e194c7762adaefb952555a/modules/imgcodecs/src/grfmt_exr.cpp#L697-L700

Reply:
I reconsider it. I fixed it.

It is better that this fallback do not be processed if selected encoder doesn't supports CV_8U (Maybe only OpenEXR encoder).